### ISSUE TITLE
Support `${...}` template expressions in attribute values

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform-ls#522]
Depends on https://github.com/hashicorp/terraform-ls/issues/496

--- 

## Background

While in many cases expressions can be declared directly after `=`, e.g. `attribute = max(1, 4)` or `attr = var.test`, sometimes more complex expression combinations require use of so called "template expression" with the interpolation syntax.

For example:

```hcl
message = "Hello, ${var.name}!"
```

Aside from interpolation syntax (`${}`) Terraform also supports string directives, as documented at https://www.terraform.io/docs/language/expressions/strings.html#directives

This allows users to write e.g. the following:

```hcl
"Hello, %{ if var.name != "" }${var.name}%{ else }unnamed%{ endif }!"
```

These directives are not supported in LS today and so they don't receive any completion, hover, or highlighting (beyond what client provides via static grammar).

Users would benefit from having the template expression supported throughout the LS functions.

### Completion

![Screenshot 2022-06-13 at 21 54 11](https://user-images.githubusercontent.com/287584/173443729-a97c386c-8840-4020-8d5d-34419facaf95.png)

![Screenshot 2022-06-13 at 22 03 46](https://user-images.githubusercontent.com/287584/173445341-602b4694-e94c-479b-a17a-d3e879658fbf.png)

### Hover

![Screenshot 2022-06-13 at 21 59 37](https://user-images.githubusercontent.com/287584/173444606-a19e9c59-fbaf-4d41-8dc4-cb1e33fb0352.png)

![Screenshot 2022-06-13 at 22 04 51](https://user-images.githubusercontent.com/287584/173445344-113d445e-54d0-40a0-9d39-413ab9ea3998.png)

### Semantic Tokens

```hcl
variable "example" {
}

resource "aws_instance" "name" {
  tags = {
    Name = "instance-${var.example}"
          # ^^^^^^^^^ hcl-string
                     # ^^^ hcl-traversalStep
                         # ^^^^^^^ hcl-traversalStep
  }
}
```

```hcl
variable "name" {
}

resource "aws_instance" "name" {
  tags = {
    Name = "Hello %{if var.name != ""}${var.name}%{else}unnamed%{endif}!"
          # ^^^^^^ hcl-string
                     # ^^^ hcl-traversalStep
                         # ^^^^ hcl-traversalStep
                                      # ^^^ hcl-traversalStep
                                          # ^^^^ hcl-traversalStep
                                                      # ^^^^^^^ hcl-string
  }
}
```

### `hcl-lang`

 - [x] Reflect `TemplateExpr` in [`CompletionAtPos()`](https://pkg.go.dev/github.com/hashicorp/hcl-lang@v0.0.0-20220610083213-029528cb7aa7/decoder#PathDecoder.CandidatesAtPos)
 - [x] Reflect `TemplateExpr` in [`HoverAtPos()`](https://pkg.go.dev/github.com/hashicorp/hcl-lang@v0.0.0-20220610083213-029528cb7aa7/decoder#PathDecoder.HoverAtPos)
 - [x] Reflect `TemplateExpr` in [`SemanticTokensInFile()`](https://pkg.go.dev/github.com/hashicorp/hcl-lang@v0.0.0-20220610083213-029528cb7aa7/decoder#PathDecoder.SemanticTokensInFile)

### `terraform-ls`

 - [x] Bump `hcl-lang` to include ^

**Comment by RobinFrcd** [comment]
Hi,
I see this ticket has been in multiple milestones, and that the dependency (https://github.com/hashicorp/terraform-ls/issues/496) has been fixed in April.
Is there any plan to include this in a future milestones ?

Thank you !

**Comment by dbanck** [comment]
@RobinFrcd We just released a new version [2.29.1](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.29.1). This version includes support for template expressions (but without directives). The update should appear automatically in VS Code.

**Comment by radeksimko** [comment]
As mentioned, this was implemented as part of https://github.com/hashicorp/hcl-lang/pull/322 and conditional directives were also implemented in https://github.com/hashicorp/hcl-lang/pull/326

For loop directive is tracked under https://github.com/hashicorp/terraform-ls/issues/527
